crates/rest-service/src/endpoints/call_contract_offchain_data.rs

```diff
@@ -0,0 +1,181 @@
+//! Endpoint for calling a contract with offchain data.
+use std::sync::Arc;
+
+use amplifier_api::types::{
+    CallEvent, CallEventMetadata, Event, EventBase, EventId, EventMetadata, GatewayV2Message,
+    MessageId, PublishEventsRequest, TxId,
+};
+use axelar_solana_gateway::processor::{CallContractOffchainDataEvent, GatewayEvent};
+use axum::extract::{Json, State};
+use axum::http::StatusCode;
+use axum::response::IntoResponse;
+use axum::routing::{post, MethodRouter};
+use futures::channel::mpsc::UnboundedSender;
+use futures::SinkExt as _;
+use gateway_event_stack::{MatchContext, ProgramInvocationState};
+use relayer_amplifier_api_integration::AmplifierCommand;
+use serde::{Deserialize, Serialize};
+use solana_client::nonblocking::rpc_client::RpcClient;
+use solana_listener::{fetch_logs, SolanaTransaction};
+use solana_sdk::signature::Signature;
+use thiserror::Error;
+
+use crate::component::ServiceState;
+
+pub(crate) const PATH: &str = "/call-contract-offchain-data";
+pub(crate) fn handlers() -> MethodRouter<Arc<ServiceState>> {
+    post(post_handler)
+}
+
+/// Payload for the call-contract-offchain-data endpoint.
+#[derive(Serialize, Deserialize)]
+pub struct Payload {
+    /// Signature of the transaction to fetch the event from.
+    pub signature: Signature,
+    /// The data to send to the destination contract.
+    pub data: String,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum CallContractOffchainDataError {
+    /// Failed to fetch transaction logs.
+    #[error("Failed to fetch transaction logs: {0}")]
+    FailedToFetchTransactionLogs(#[from] eyre::Report),
+
+    /// The hash of the given payload doesn't match the one emitted in the transaction logs.
+    #[error("Payload hashes don't match")]
+    PayloadHashMismatch,
+
+    #[error("Successful transaction with CallContractOffchainDataEvent not found")]
+    EventNotFound,
+
+    #[error("Failed to relay message to Axelar Amplifier")]
+    FailedToRelayToAmplifier,
+
+    #[error("Failed to decode base58 data: {0}")]
+    DecodeError(#[from] bs58::decode::Error),
+}
+
+impl IntoResponse for CallContractOffchainDataError {
+    fn into_response(self) -> axum::response::Response {
+        let error_tuple = match self {
+            Self::FailedToFetchTransactionLogs(report) => {
+                (StatusCode::INTERNAL_SERVER_ERROR, report.to_string())
+            }
+            Self::FailedToRelayToAmplifier | Self::DecodeError(_) => {
+                (StatusCode::INTERNAL_SERVER_ERROR, self.to_string())
+            }
+
+            Self::PayloadHashMismatch => (StatusCode::BAD_REQUEST, self.to_string()),
+            Self::EventNotFound => (StatusCode::NOT_FOUND, self.to_string()),
+        };
+
+        error_tuple.into_response()
+    }
+}
+
+async fn post_handler(
+    State(state): State<Arc<ServiceState>>,
+    Json(payload): Json<Payload>,
+) -> Result<(), CallContractOffchainDataError> {
```

Comment:
misc:
add some logging.
How I'd approach it:
- create a span that captures the signature from the request params
- emit an `info!` log if tx was found and succesfully forwarded, `warn!` log otherwise

Reply:
Looks great 🚀